examples/bridge-app/esp32/main/main.cpp

```diff
@@ -341,7 +341,7 @@ static void InitServer(intptr_t context)
     // Set starting endpoint id where dynamic endpoints will be assigned, which
     // will be the next consecutive endpoint id after the last fixed endpoint.
     gFirstDynamicEndpointId = static_cast<chip::EndpointId>(
```

Comment:
So if there is only one fixed endpoint (ep0), then `emberAfFixedEndpointCount()` is going to be 1, we'll do `emberAfEndpointFromIndex(0)`, that will return 0, and after this change we will set `gFirstDynamicEndpointId` to 0, right?

That seems a little odd.

Reply:
I checked again, in the unmodified version I get the `gFirstDynamicEndpointId` as `3`.
```
I (2176) chip[DL]: gFirstDynamicEndpointId = 3
I (2176) chip[DL]: Added device Light 1 to dynamic endpoint 3 (index=0)
I (2186) chip[DL]: gFirstDynamicEndpointId = 3
I (2196) chip[DL]: Added device Light 2 to dynamic endpoint 4 (index=1)
I (2196) chip[DL]: gFirstDynamicEndpointId = 3
I (2206) chip[DL]: Added device Light 3 to dynamic endpoint 5 (index=2)
I (2216) chip[DL]: Removed device Light 2 from dynamic endpoint 4 (index=1)
I (2216) chip[DL]: gFirstDynamicEndpointId = 3
I (2226) chip[DL]: Added device Light 4 to dynamic endpoint 6 (index=1)
I (2236) chip[DL]: gFirstDynamicEndpointId = 3
I (2236) chip[DL]: Added device Light 2 to dynamic endpoint 7 (index=3)
```

I just logged the `gFirstDynamicEndpointId` as the first line of the function: https://github.com/project-chip/connectedhomeip/blob/7fde831b69d70785ff5d8116dba31a0213da23dd/examples/bridge-app/esp32/main/main.cpp#L143-L178.

But I would have expected `gFirstDynamicEndpointId` to be `2` therefore I suggested in this PR to remove the `+ 1` from the line:
https://github.com/project-chip/connectedhomeip/blob/7fde831b69d70785ff5d8116dba31a0213da23dd/examples/bridge-app/esp32/main/main.cpp#L343-L344

So I understand EP0 is the reserved endpoint and EP1 is the one with the clusters for the bridge itself and the all virtual lights should start from `2` not `3`.

Or do I get something wrong here? But the spec stated something like that the bridged devices should start on the first free endpoint and increment from there.

Reply:
Ah, so the key here is that there is an extra fixed endpoint, with endpoint id 2, that is used for some bookkeeping and disabled a few lines down.  So emberAfFixedEndpointCount() is 3, we call emberAfEndpointFromIndex(2), that returns that endpoint id (2), and the first endpoint id available for dynamic endpoints is 3.  Which is what `gFirstDynamicEndpointId` gets set to.

So having the lights start at endpoint 3 is correct.

Without the +1, there are now two endpoints with endpoint id 2 (one at index 2 which is disabled, one at index 3 which is enabled, I suspect), and things get pretty confused.

Reply:
I think the solution here should be to remove the fixed endpoint (EP 2): `dimmablelight` from the zap

Reply:
Bridge has 3 fixed endpoints

0. Matter Root Node
1. Matter bridge
2. Matter Dimmable Light

If we remove dimmablelight from zap, compilation fails. Though endpoint 0 and 1 don’t have level-control and on-off cluster
[logs.txt](https://github.com/project-chip/connectedhomeip/files/9156319/logs.txt)
.

Reply:
Ohh, may be that's the reason it was added?

Reply:
What confuses me is that we do not have EP2 for the Matter Dimmable Light in the unmodified version. If I try to access the endpoint for the descriptor cluster I can't access it as all and get an `UNSUPPORTED_ENDPOINT` error.

```
./chip-tool descriptor read parts-list 12344321 2
[...]
[1658401669.157036][81959:81967] CHIP:TOO: Response Failure: IM Error 0x0000057F: General error: 0x7f (UNSUPPORTED_ENDPOINT)
```

In contrast EP0 and EP3 result in the expected output:
```
./chip-tool descriptor read parts-list 12344321 0
[...]
[1658403531.046090][107132:107137] CHIP:TOO: Endpoint: 0 Cluster: 0x0000_001D Attribute 0x0000_0003 DataVersion: 1949639396
[1658403531.046110][107132:107137] CHIP:TOO:   parts list: 5 entries
[1658403531.046115][107132:107137] CHIP:TOO:     [1]: 1
[1658403531.046119][107132:107137] CHIP:TOO:     [2]: 3
[1658403531.046122][107132:107137] CHIP:TOO:     [3]: 6
[1658403531.046125][107132:107137] CHIP:TOO:     [4]: 5
[1658403531.046128][107132:107137] CHIP:TOO:     [5]: 7
```

```
./chip-tool descriptor read parts-list 12344321 3
[...]
[1658403575.274690][108807:108812] CHIP:TOO: Endpoint: 3 Cluster: 0x0000_001D Attribute 0x0000_0003 DataVersion: 895543079
[1658403575.274705][108807:108812] CHIP:TOO:   parts list: 0 entries

```

Reply:
@Tim-Conrad EP2 is there in the configuration, but disabled at runtime.  It's a dummy endpoint used for bookkeeping purposes (specifically to trigger code-generation of the right dispatch code for command handling in ZAP).

Reply:
@bzbarsky-apple Okay understood, so this is a workaround for the dynamic endpoint generation and not a bug/error in the code? So at least the README should be updated to show the correct endpoints

https://github.com/project-chip/connectedhomeip/blob/6d5480788b783c6cf80bdb019476971b5c5c8fb0/examples/bridge-app/esp32/README.md?plain=1#L8-L11

and the comments should be changed to the correct endpoints as well, right?

https://github.com/project-chip/connectedhomeip/blob/6d5480788b783c6cf80bdb019476971b5c5c8fb0/examples/bridge-app/esp32/main/main.cpp#L357-L374

Because all need to be one higher because of this removed endpoint.

Reply:
Linux bridge app would behave similarly!!